monkestation/code/modules/replays/subsystem/replay.dm

```diff
@@ -162,104 +118,149 @@ SUBSYSTEM_DEF(demo)
-			M.demo_last_appearance = M.appearance
+			appearance_string = encode_appearance(appearance, target = M)
+			M.demo_last_appearance = appearance
+		else if(appearance != demo_last_appearance)
+			appearance_string = encode_appearance(appearance, demo_last_appearance, TRUE, M)
+			M.demo_last_appearance = appearance
 		dirty_updates += "\ref[M] [loc_string] [appearance_string]"
 		if(MC_TICK_CHECK)
 			canceled = TRUE
 			break
 	if(length(dirty_updates))
-		var/s = "update [jointext(dirty_updates, ",")]\n"
-		WRITE_LOG_NO_FORMAT(GLOB.demo_log, s)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "update [jointext(dirty_updates, ",")]\n")
 	if(canceled)
 		return
 
-
-	var/list/marked_new = src.marked_new
 	var/list/new_updates = list()
 	while(length(marked_new))
-		last_completed++
+		// last_completed++
 		var/atom/movable/M = marked_new[length(marked_new)]
 		marked_new.len--
 		if(QDELETED(M))
 			continue
 		var/loc_string = "null"
-		if(isturf(M.loc))
+		var/loc = M.loc
+		var/appearance = M.appearance
+		if(isturf(loc))
 			loc_string = "[M.x],[M.y],[M.z]"
-		else if(ismovable(M.loc))
-			loc_string = "\ref[M.loc]"
-		M.demo_last_appearance = M.appearance
-		new_updates += "\ref[M] [loc_string] [encode_appearance(M.appearance, target = M)]"
+		else if(ismovable(loc))
+			loc_string = ref(loc)
+		M.demo_last_appearance = appearance
+		new_updates += "\ref[M] [loc_string] [encode_appearance(appearance, target = M)]"
 		if(MC_TICK_CHECK)
 			canceled = TRUE
 			break
 	if(length(new_updates))
-		var/s = "new [jointext(new_updates, ",")]\n"
-		WRITE_LOG_NO_FORMAT(GLOB.demo_log, s)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "new [jointext(new_updates, ",")]\n")
 	if(canceled)
 		return
 
-
-	var/list/marked_turfs = src.marked_turfs
 	var/list/turf_updates = list()
 	while(length(marked_turfs))
-		last_completed++
+		// last_completed++
 		var/turf/T = marked_turfs[length(marked_turfs)]
 		marked_turfs.len--
-		if(T && T.appearance != T.demo_last_appearance)
-			turf_updates += "([T.x],[T.y],[T.z])=[encode_appearance(T.appearance, T.demo_last_appearance)]"
-			T.demo_last_appearance = T.appearance
+		if(isnull(T))
+			continue
+		var/appearance = T.appearance
+		var/demo_last_appearance = T.demo_last_appearance
+		if(appearance != demo_last_appearance)
+			turf_updates += "([T.x],[T.y],[T.z])=[encode_appearance(appearance, demo_last_appearance)]"
+			T.demo_last_appearance = appearance
 			if(MC_TICK_CHECK)
 				canceled = TRUE
 				break
 	if(length(turf_updates))
-		var/s = "turf [jointext(turf_updates, ",")]\n"
-		WRITE_LOG_NO_FORMAT(GLOB.demo_log, s)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "turf [jointext(turf_updates, ",")]\n")
 	if(canceled)
 		return
 
+/datum/controller/subsystem/demo/stat_entry(msg)
+	msg += "Remaining: {"
+	msg += "Trf:[length(marked_turfs)]|"
+	msg += "New:[length(marked_new)]|"
+	msg += "Upd:[length(marked_dirty)]|"
+	msg += "Del:[length(del_list)]"
+	msg += "}"
+	return ..()
+
+/datum/controller/subsystem/demo/proc/disable()
```

Comment:
moved to its own proc to avoid code repetition in case it needs to be disabled outside of init